pulsar/src/main/java/com/hazelcast/jet/contrib/pulsar/PulsarSources.java

```diff
@@ -0,0 +1,374 @@
+     * @param projectionFn converts a Pulsar message to an emitted item.
+     * @param <M> the type of the message read by {@code PulsarConsumer}
+     * @param <T> the type of data emitted from {@code StreamSource}
+     * @return a stream source to use in {@link Pipeline#readFrom}
+     */
+    public static <M, T> StreamSource<T> pulsarConsumer(
+            @Nonnull List<String> topics,
+            @Nonnull int preferredLocalParallelism,
+            @Nonnull Map<String, Object> consumerConfig,
+            @Nonnull SupplierEx<PulsarClient> connectionSupplier,
+            @Nonnull SupplierEx<Schema<M>> schemaSupplier,
+            @Nonnull FunctionEx<Message<M>, T> projectionFn
+    ) {
+        checkSerializable(connectionSupplier, "connectionSupplier");
+        return SourceBuilder.timestampedStream("pulsar-distributed-stream-source", ctx -> new ConsumerContext(
+                ctx.logger(), connectionSupplier.get(), topics, consumerConfig, schemaSupplier, projectionFn))
+                .<T>fillBufferFn(ConsumerContext::fillBuffer)
+                .destroyFn(ConsumerContext::destroy)
+                .distributed(Vertex.checkLocalParallelism(preferredLocalParallelism))
+                .build();
+    }
+
+
+    /**
+     * Creates a fault-tolerant timestamped {@link StreamSource}
+     * which reads messages from Pulsar topics for data ingestion to Jet
+     * pipelines. It uses the Reader API of the Pulsar client. It cannot
+     * be used in the partitioned topics.
+     * <p>
+     * Example usage:
+     * <pre>{@code
+     *      Properties credentials = loadTwitterCredentials();
+     *      TimestampedS<Status> twitterSearchSource =
+     *      TwitterSources.search(credentials,"Jet flies");
+     *      Pipeline p = Pipeline.create();
+     *      BatchStage<Status> srcStage = p.readFrom(pulsarSource);
+     *  }</pre>
+     *
+     * @param topic          the single topic to consume
+     * @param readerConfig   Pulsar reader configurations that must contain
+     *                       reader name
+     * @param schemaSupplier supplies the schema for consuming messages
+     * @param projectionFn   converts a Pulsar message to an emitted item.
+     * @param <M>            the type of the message read by {@code PulsarConsumer}
+     * @param <T>            the type of data emitted from {@code StreamSource}
+     * @return a stream source to use in {@link Pipeline#readFrom}
+     */
+
+    public static <M, T> StreamSource<T> pulsarReader(
+            @Nonnull String topic,
+            @Nonnull Map<String, Object> readerConfig,
+            @Nonnull SupplierEx<PulsarClient> connectionSupplier,
+            @Nonnull SupplierEx<Schema<M>> schemaSupplier,
+            @Nonnull FunctionEx<Message<M>, T> projectionFn
+    ) {
+        checkSerializable(connectionSupplier, "connectionSupplier");
+        return SourceBuilder.timestampedStream("pulsar-ft-stream-source", ctx -> new ReaderContext(
```

Comment:
`new ReaderContext` -> `new ReaderContext<>`

Reply:
Changed this.

Reply:
fixed.